### ISSUE TITLE
Breaking glibc changes cause crashes due to impurities

### ISSUE THREAD

**Issue by primeos** [NixOS/nixpkgs#95808]
## Issue description

This is a meta issue to track this problem (I'm not aware of an existing issue). I'm not sure if we can do much here but it seems like the `glibc` library is often loaded from another location that the hard-coded Nix store path (I haven't looked into the why and how).

### Related issues

- https://github.com/NixOS/nixpkgs/issues/94315
- https://github.com/NixOS/nixpkgs/issues/95425
- https://github.com/NixOS/nixpkgs/issues/95769
- https://github.com/NixOS/nixpkgs/issues/95806

**Comment by vcunat** [comment]
Real libGL is loaded impurely, by design.  I expect the problems are usually caused by that, and they should be gone by switching NixOS to a similar version (or using something like nixGL in use cases where that seems better).

**Comment by primeos** [comment]
> Real libGL is loaded impurely, by design. I expect the problems are usually caused by that

Yeah me too. Though I have never looked into how this actually works (and I'm lacking a lot of knowledge in this area). In the end both the application and libGL should refer to the correct `glibc` version. IIRC in one of the linked issues it was mentioned that the `glibc` version from the application might get reused for libGL (which sounds plausible since libGL is also loaded by the application - though libglvnd AFAIK, but I also never looked at that). If that's what's happening ("library reuse") then I guess we cannot do anything here (as noted above).

But https://github.com/NixOS/nixpkgs/issues/95425 and https://github.com/NixOS/nixpkgs/issues/95769 aren't caused by the Mesa drivers. https://github.com/NixOS/nixpkgs/issues/95769 could be similar (differing Nixpkgs revisions for VA-API drivers and the client application) but https://github.com/NixOS/nixpkgs/issues/95425 might be a bit different (though also two differing Nixpkgs revisions, but this is obviously also a requirement to trigger this issue in the first place).

Basically I wanted to open this issue to investigate if we can somehow load a different `glibc` for a library dependency of an executable (which might not be possible, I have no idea here). If that isn't possible then I guess we cannot do anything about such issues.

**Comment by vcunat** [comment]
The libraries gets reused and **that** is what causes the breakage, I believe.  If you compile against some glibc version, you can run against any higher one &ndash; there's an insane level of compatibility in there &ndash; but not the other way around.  IIRC some libGLs also use other libraries like libstdc++ and even some libwayland-*.

Well, better look at older threads like https://github.com/NixOS/nixpkgs/issues/31189

**Comment by primeos** [comment]
> Basically I wanted to open this issue to investigate if we can somehow load a different glibc for a library dependency of an executable (which might not be possible, I have no idea here). If that isn't possible then I guess we cannot do anything about such issues.

I did some basic research on this and unfortunately this doesn't seem possible in the way that I hoped. Our problem here seems to be called a "[diamond dependency](https://en.wikipedia.org/wiki/Dependency_hell#Problems)".

I guess linking `glibc` statically should work (`-Wl,-Bstatic -libc -Wl,-Bdynamic`?), but that would only make sense for specific packages like Mesa that often cause such conflicts. Similarly `libcapsule` and `dlmopen()` (from #31189) also sound interesting though that should also only make sense for specific packages (and doesn't seem to fully work for our use-case yet - though https://gitlab.collabora.com/vivek/libcapsule seems at least to be actively maintained).

> The libraries gets reused and that is what causes the breakage, I believe.

I didn't verify it but that surely seems to be the case.

> If you compile against some glibc version, you can run against any higher one – there's an insane level of compatibility in there – but not the other way around.

Good to know, I was surprised as I thought `glibc` wouldn't even be backward compatible but it turns out that the reported functions where only moved, not removed ([source](https://sourceware.org/legacy-ml/libc-alpha/2019-08/msg00029.html)):
> * The functions clock_gettime, clock_getres, clock_settime,
  clock_getcpuclockid, clock_nanosleep were removed from the librt library
  for new applications (on architectures which had them).  Instead, the
  definitions in libc will be used automatically, which have been available
  since glibc 2.17.

Though there often are a few breaking changes: [API/ABI changes review for glibc](https://abi-laboratory.pro/?view=timeline&l=glibc)

> IIRC some libGLs also use other libraries like libstdc++ and even some libwayland-*.

Indeed. #95806 is caused by libstdc++. That unfortunately makes everything more complicated... :o

> Well, better look at older threads like #31189

Do you think I should close this issue? My intention was to find a general solution for the diamond dependency problem (at least for `glibc`), rather than a solution for the problems introduced though `/run/opengl-driver/`. But since a suitable solution for this issue might not even exist we can also close it.

**Comment by vcunat** [comment]
:shrug:  Almost all such problems I've seen around nixos.org apparently do come from `/run/opengl-driver`, so improving that would most likely be helpful to some people.  We don't have many other library impurities, I think.

**Comment by doronbehar** [comment]
Is https://github.com/NixOS/nixpkgs/issues/101457 an example of a similar glibc issue?

**Comment by xaverdh** [comment]
Sorry if this is a naive suggestion, but could this be resolved by statically linking everything (e.g. glibc) in the drivers themselves?

**Comment by gebner** [comment]
>  Almost all such problems I've seen around nixos.org apparently do come from `/run/opengl-driver`, so improving that would most likely be helpful to some people. We don't have many other library impurities, I think.

Two other common ones are:
 1. Input methods (recent glib bump broke input method support in applications linked to older glib version).
 2. Qt plugins such as themes (see recent flood of issues about KDE themes (built against 5.12) not working with Qt apps (built against 5.15).)

Neither of them cause segfaults though, AFAICT.

**Comment by doronbehar** [comment]
Another example of a niche impurity is the sane nixos module which adds a LD_LIBRARY_PATH to the environment of all users, where only 1 or 2 packages use it in practice:

https://github.com/NixOS/nixpkgs/blob/7e25b7113271cf80eb0f950da2e1aa7b382eb268/nixos/modules/services/hardware/sane.nix#L29-L32

**Comment by evils** [comment]
i'm not sure this is the right place to post this
but i've not been able to develop a rust firmware project of mine for a while
`-glibc-2.31/lib/libc.so.6: version 'GLIBC_2.32' not found`

i've got a minimal reproduction set up here, just run `nix-shell --command 'cargo build'`
https://gitlab.com/evils/rust-glibc-test

**Comment by evils** [comment]
correction to the above, full command to reproduce the error is  
`git clone https://gitlab.com/evils/rust-glibc-test && cd rust-glibc-test && git checkout dd3f2bf8 && nix-shell --command 'rustup install stable && rustup default stable && rustup target add thumbv6m-none-eabi && cargo update && cargo clean && cargo build && exit'`

this still results in the error on nixos-unstable on x86_64  
however it does not result in a failure on aarch64

i'm currently using nixos-20.03 for my shell  
anyone know if there's hope of this error going away?

**Comment by doronbehar** [comment]
I'm no rust developer but I experienced issues with rustup executables in the past which may be related:

https://discourse.nixos.org/t/do-rustup-installed-components-such-as-rls-supposed-to-be-usable-on-nixos/6728

**Comment by evils** [comment]
doh, running `rustup update` resolved my issue  
i had tried removing ~/.cargo, forgot about ~/.rustup  

thanks for the hint
